src/pycsou/opt/solver/pds.py

```diff
@@ -0,0 +1,965 @@
+            u = 2 * x_temp - mst["x"]
+            z_temp = self._h.fenchel_prox(mst["z"] + mst["sigma"] * self._K(u), sigma=mst["sigma"])
+            mst["z"] = mst["rho"] * z_temp + (1 - mst["rho"]) * mst["z"]
+        mst["x"] = mst["rho"] * x_temp + (1 - mst["rho"]) * mst["x"]
+
+    def default_stop_crit(self) -> pycs.StoppingCriterion:
+        stop_crit_x = pycos.RelError(
+            eps=1e-4,
+            var="x",
+            f=None,
+            norm=2,
+            satisfy_all=True,
+        )
+        stop_crit_z = pycos.RelError(
+            eps=1e-4,
+            var="z",
+            f=None,
+            norm=2,
+            satisfy_all=True,
+        )
+        return stop_crit_x & stop_crit_z
+
+    def solution(self) -> pyct.NDArray:
+        """
+        Returns
+        -------
+        p: NDArray
+            (..., N) x solution.
+        d: NDArray
+            (..., N) z solution.
+        """
+        data, _ = self.stats()
+        return data.get("x")
+
+    def set_beta(self, beta: typ.Optional[pyct.Real]) -> float:
+        r"""
+        Sets the Lipschitz constant.
+        Returns
+        -------
+        float
+            Lipschitz constant.
+        """
+        if beta is None:
+            if math.isfinite(dl := self._f._diff_lipschitz):
+                return pycrt.coerce(dl)
+            else:
+                msg = "beta: automatic inference not supported for operators with unbounded Lipschitz gradients."
+            raise ValueError(msg)
+        else:
+            try:
+                assert beta > 0
+                return pycrt.coerce(beta)
+            except:
+                raise ValueError(f"beta must be positive, got {beta}.")
+
+    def set_step_sizes(self, tau: typ.Optional[pyct.Real], sigma: typ.Optional[pyct.Real]) -> typ.Tuple[float, float]:
+        r"""
+        Set the x/z step sizes.
+        Returns
+        -------
+        Tuple[float, float]
+            Sensible x/z step sizes.
+        Notes
+        -----
+        In practice, the convergence speed  of the algorithm is improved by choosing :math:`\sigma` and :math:`\tau` as large as possible and relatively well-balanced --so that both the x and z variables converge at the same pace. In practice, it is hence recommended to choose perfectly balanced parameters :math:`\sigma=\tau` saturating the convergence inequalities.
+        For :math:`\beta>0` this yields:
+        .. math::
+           \frac{1}{\tau}-\tau\Vert\mathbf{K}\Vert_{2}^2= \frac{\beta}{2} \quad\Longleftrightarrow\quad -2\tau^2\Vert\mathbf{K}\Vert_{2}^2-\beta\tau+2=0,
+        which admits one positive root
+        .. math::
+           \tau=\sigma=\frac{1}{\Vert\mathbf{K}\Vert_{2}^2}\left(-\frac{\beta}{4}+\sqrt{\frac{\beta^2}{16}+\Vert\mathbf{K}\Vert_{2}^2}\right).
+        For :math:`\beta=0`, this yields
+        .. math::
+           \tau=\sigma=\Vert\mathbf{K\Vert_{2}^{-1}.}
+        """
+
+        if (tau is not None) and (sigma is None):
```

Comment:
The first two ``if`` cases are wrong: one cannot just set ``tau=sigma`` for any value of ``tau/sigma``. We should discuss that.